server/finatra-server/src/main/scala/tapir/server/finatra/OutputToFinatraResponse.scala

```diff
@@ -0,0 +1,141 @@
+            vs(i).asInstanceOf[Seq[(String, String)]].foldLeft(finatraResponse) {
+              case (fr, (name, value)) => fr.setOrAddHeader(name, value)
+            }
+
+          case (EndpointIO.Mapped(wrapped, _, g, _), i) =>
+            apply(wrapped, g(vs(i)), Some(finatraResponse))
+
+          case (EndpointOutput.StatusCode(), i) =>
+            finatraResponse.copy(status = Status(vs(i).asInstanceOf[Int]))
+
+          case (EndpointOutput.StatusFrom(io, default, _, when), i) =>
+            val v = vs(i)
+            val sc = when.find(_._1.matches(v)).map(_._2).getOrElse(default)
+            apply(io, v).copy(status = Status(sc))
+
+          case (EndpointOutput.Mapped(wrapped, _, g, _), i) =>
+            apply(wrapped, g(vs(i)), Some(finatraResponse))
+        }
+    }
+
+  }
+
+  private def rawValueToFinatraContent[M <: MediaType, R](codecMeta: CodecMeta[M, R], r: R): (FinatraContent, String) = {
+    val ct: String = codecMeta.mediaType.mediaType
+
+    codecMeta.rawValueType match {
+      case StringValueType(charset) =>
+        FinatraContentBuf(Buf.ByteArray.Owned(r.toString.getBytes(charset))) -> ct
+      case ByteArrayValueType  => FinatraContentBuf(Buf.ByteArray.Owned(r)) -> ct
+      case ByteBufferValueType => FinatraContentBuf(Buf.ByteBuffer.Owned(r)) -> ct
+      case InputStreamValueType =>
+        FinatraContentReader(Reader.fromStream(r: InputStream)) -> ct
+      case FileValueType =>
+        FinatraContentReader(Reader.fromFile(r: File)) -> ct
+      case mvt: MultipartValueType =>
+        val entity = MultipartEntityBuilder.create()
+
+        (r: Seq[RawPart]).flatMap(rawPartToFormBodyPart(mvt, _)).foreach { formBodyPart: FormBodyPart =>
+          entity.addPart(formBodyPart)
+        }
+
+        FinatraContentReader(InputStreamReader(entity.build().getContent)) -> ct
+    }
+  }
+
+  private def rawValueToContentBody[M <: MediaType, R](codecMeta: CodecMeta[M, R], part: Part[R], r: R): ContentBody = {
+    val contentType: String = part.header("content-type").getOrElse("text/plain")
+
+    codecMeta.rawValueType match {
+      case StringValueType(charset) =>
+        new StringBody(r.toString, ContentType.create(contentType, charset))
+      case ByteArrayValueType =>
+        new ByteArrayBody(r: Array[Byte], ContentType.create(contentType), part.fileName.get)
+      case ByteBufferValueType =>
+        val array: Array[Byte] = new Array[Byte]((r: ByteBuffer).remaining)
+        (r: ByteBuffer).get(array)
+        new ByteArrayBody(array, ContentType.create(contentType), part.fileName.get)
+      case FileValueType =>
+        part.fileName match {
+          case Some(filename) => new FileBody(r: File, ContentType.create(contentType), filename)
+          case None           => new FileBody(r: File, ContentType.create(contentType))
+        }
+      case InputStreamValueType =>
+        new InputStreamBody(r: InputStream, ContentType.create(contentType), part.fileName.get)
+      case _ =>
```

Comment:
is that for multiparts? Do I correctly understand that a multipart here would mean that it's nested? Maybe we can throw a meaningful exception here then (like "nested multiparts are not supported in output X")

Reply:
Oops yeah I had that catch-all when developing and missed that case. I believe the only thing left here is multiparts, yeah. I'll update that to catch the specific `MultipartValueType` here-- Is just throwing here appropriate or do we need to return an appropriate Try/Either so that it's handled correctly by the error-handling code?